sample-libs/Samples.DatabaseHelper/RelationalDatabaseTestHarness.cs

```diff
@@ -1,303 +1,157 @@
-        private const string InsertCommandText = "INSERT INTO Employees (Id, Name) VALUES (@Id, @Name);";
-        private const string SelectOneCommandText = "SELECT Name FROM Employees WHERE Id=@Id;";
-        private const string UpdateCommandText = "UPDATE Employees SET Name=@Name WHERE Id=@Id;";
-        private const string SelectManyCommandText = "SELECT * FROM Employees WHERE Id=@Id;";
-        private const string DeleteCommandText = "DELETE FROM Employees WHERE Id=@Id;";
-
-        private readonly TConnection _connection;
-
-        private readonly Func<TCommand, int> _executeNonQuery;
-        private readonly Func<TCommand, object> _executeScalar;
-        private readonly Func<TCommand, TDataReader> _executeReader;
-        private readonly Func<TCommand, CommandBehavior, TDataReader> _executeReaderWithBehavior;
-
-        private readonly Func<TCommand, Task<int>> _executeNonQueryAsync;
-        private readonly Func<TCommand, Task<object>> _executeScalarAsync;
-        private readonly Func<TCommand, Task<TDataReader>> _executeReaderAsync;
-        private readonly Func<TCommand, CommandBehavior, Task<TDataReader>> _executeReaderWithBehaviorAsync;
-
-        public RelationalDatabaseTestHarness(
-            TConnection connection,
-            Func<TCommand, int> executeNonQuery,
-            Func<TCommand, object> executeScalar,
-            Func<TCommand, TDataReader> executeReader,
-            Func<TCommand, CommandBehavior, TDataReader> executeReaderWithBehavior,
-            Func<TCommand, Task<int>> executeNonQueryAsync,
-            Func<TCommand, Task<object>> executeScalarAsync,
-            Func<TCommand, Task<TDataReader>> executeReaderAsync,
-            Func<TCommand, CommandBehavior, Task<TDataReader>> executeReaderWithBehaviorAsync)
-        {
-            _connection = connection ?? throw new ArgumentNullException(nameof(connection));
-
-            _executeNonQuery = executeNonQuery ?? throw new ArgumentNullException(nameof(executeNonQuery));
-            _executeScalar = executeScalar ?? throw new ArgumentNullException(nameof(executeScalar));
-            _executeReader = executeReader ?? throw new ArgumentNullException(nameof(executeReader));
-            _executeReaderWithBehavior = executeReaderWithBehavior ?? throw new ArgumentNullException(nameof(executeReaderWithBehavior));
-
-            // async methods are not implemented by all ADO.NET providers, so they can be null
-            _executeNonQueryAsync = executeNonQueryAsync;
-            _executeScalarAsync = executeScalarAsync;
-            _executeReaderAsync = executeReaderAsync;
-            _executeReaderWithBehaviorAsync = executeReaderWithBehaviorAsync;
-        }
-
-        public async Task RunAsync()
+        public static async Task RunAllAsync(
+            IDbConnection connection,
+            DbCommandFactory commandFactory,
+            IDbCommandExecutor commandExecutor,
```

Comment:
nit: On my first read I had a hard time understanding what `commandExecutor` was supposed to represent. Is this variable always supposed to represent a different library's DbCommand implementation? If so, could you rename this to make that intent more clear?

Reply:
Yes, this `commandExecutor` is for the specific ADO.NET provider (`SqlCommand`, `NpgsqlCommand, `MySqlCommand`, etc) and must be provided by the sample app. We can create the others (`DbCommand, `IDbCommand`) ourselves.

Re: naming, I keep thinking "executioner", haha. It was all I came up with for something that "executes [methods in] DbCommands". Open to suggestions!

Reply:
Renamed and added lots of comments in 6de45067

Reply:
Sounds good, I was trying to think of suggestions and couldn't 😄